### PR TITLE
WD-8625 - feat: Display errors in WebCLI

### PR DESCRIPTION
**Pull request by vladimir-cucu**
## Done

- Display connection and authentication errors in WebCLI

## Details

- https://warthogs.atlassian.net/browse/WD-8625

### CODE REVIEW

File: src/components/WebCLI/WebCLI.tsx [hunk 1]
```diff
@@ -41,6 +46,13 @@ const WebCLI = ({
   const [connection, setConnection] = useState<Connection | null>(null);
   const [shouldShowHelp, setShouldShowHelp] = useState(false);
   const [historyPosition, setHistoryPosition] = useState(0);
+  // First element in inLineErrors array corresponds to the error when no
+  // websocket address is provided. Second element corresponds to the error
+  // when no authentication information is available for Web CLI.
+  const [inlineErrors, setInlineErrors] = useState<(string | null)[]>([
+    null,
+    null,
+  ]);
```
Comment (vladimir-cucu) [inline src/components/WebCLI/WebCLI.tsx:55]: I will change this part of the code to use the hook in your PR once it lands.

File: src/hooks/useInlineErrors.ts [hunk 2]
```diff
@@ -21,7 +21,8 @@ function useInlineErrors(
   const [inlineErrors, setInlineErrors] = useState<InlineError[]>([]);
   const setError = useCallback(
     (key: InlineError["key"], error: InlineError["error"]) => {
-      setInlineErrors((inlineErrors) => {
+      setInlineErrors((prevInlineErrors) => {
+        const inlineErrors = [...prevInlineErrors];
```
Comment (vladimir-cucu) [inline src/hooks/useInlineErrors.ts:25]: I think, without this fix, the Connection error won't appear straight away in the `WebCLI`, as we were mutating the state here directly.

File: src/components/WebCLI/WebCLI.tsx [hunk 3]
```diff
@@ -41,6 +46,13 @@ const WebCLI = ({
   const [connection, setConnection] = useState<Connection | null>(null);
   const [shouldShowHelp, setShouldShowHelp] = useState(false);
   const [historyPosition, setHistoryPosition] = useState(0);
+  // First element in inLineErrors array corresponds to the error when no
+  // websocket address is provided. Second element corresponds to the error
+  // when no authentication information is available for Web CLI.
+  const [inlineErrors, setInlineErrors] = useState<(string | null)[]>([
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/components/WebCLI/WebCLI.tsx]: Done

File: src/hooks/useInlineErrors.ts (hunk 2 above)
Comment (huwshimi) [reply src/hooks/useInlineErrors.ts:25]: I'd always understood updater functions to be safe as they do not mutate the current state, but rather the functions get queued to be reduced on the next render. However, I couldn't see anywhere in the documentation where it mentions whether it is safe to mutate the previous state inside the function.

I dug into the source and it looks like the updater function is just a reducer function passed to useReducer and in the docs for useReducer it says to use a new object so it doesn't have the same reference.

Saying all that, was there a specific bug that you were running into? I tried both update methods and the render order appeared to be working correctly for me.

File: src/components/WebCLI/WebCLI.tsx [hunk 4]
```diff
@@ -197,30 +211,34 @@ const WebCLI = ({
...
-              className="p-link--inverted"
-              rel="noreferrer"
-              target="_blank"
-            >
-              full documentation here
-            </a>
-            .
-          </>
+          // If errors exist, display them instead of the default help message.
+          Object.values(InlineErrors).some((error) => hasInlineError(error)) ? (
```
Comment (huwshimi) [inline src/components/WebCLI/WebCLI.tsx:222]: I think we should update `hasInlineError` to check all errors if you don't pass a key.

File: src/components/WebCLI/WebCLI.tsx [hunk 5]
```diff
@@ -197,30 +211,34 @@ const WebCLI = ({
...
+            <>
+              {inlineErrors.map((error) =>
+                error ? <div key={error as string}>ERROR: {error}</div> : null,
```
Comment (huwshimi) [inline src/components/WebCLI/WebCLI.tsx:225]: `inlineErrors` are reduced to truthy values before being returned by `useInlineErrors` so this ternary isn't necessary.

File: src/hooks/useInlineErrors.test.ts [hunk 6]
```diff
@@ -1,12 +1,12 @@
...
 import useInlineErrors from "./useInlineErrors";
 
 describe("useInlineErrors", () => {
   it("can add an error", () => {
     const { result, rerender } = renderHook(() => useInlineErrors());
     let [errors, setError] = result.current;
-    setError("Error1", "Uh oh!");
+    act(() => setError("Error1", "Uh oh!"));
     rerender();
     [errors] = result.current;
```
Comment (huwshimi) [inline src/hooks/useInlineErrors.test.ts:11]: Looks like I somehow doubled up the rerender and destructure in this test so you could remove these two lines.

File: src/hooks/useInlineErrors.test.ts [hunk 7]
```diff
@@ -17,11 +17,11 @@ describe("useInlineErrors", () => {
   it("can update an error", () => {
     const { result, rerender } = renderHook(() => useInlineErrors());
     let [errors, setError] = result.current;
-    setError("Error1", "Uh oh!");
+    act(() => setError("Error1", "Uh oh!"));
     rerender();
```
Comment (huwshimi) [inline src/hooks/useInlineErrors.test.ts:21]: Now that these are using `act()` I don't think they need to also have `rerender()`.

File: src/components/WebCLI/WebCLI.tsx [hunk 8]
```diff
@@ -159,10 +172,11 @@ const WebCLI = ({
         );
         authentication.user = activeUser ? getUserName(activeUser) : undefined;
         authentication.macaroons = [deserialized];
-      } else {
-        // XXX Surface error to the user.
-        console.error("No authentication information available");
       }
+      setInlineError(
+        InlineErrors.AUTHENTICATION,
+        macaroons ? null : Label.AUTHENTICATION_ERROR,
```
Comment (huwshimi) [inline src/components/WebCLI/WebCLI.tsx:178]: While I don't think we should pop open the console if there is a connection error, I think we should open it when someone tries to send a command and then get this error.

File: src/hooks/useInlineErrors.ts (hunk 2 above)
Comment (vladimir-cucu) [reply src/hooks/useInlineErrors.ts:25]: Thanks for the detailed explanation! After today's discussion, I looked over the code a bit and I think I'll use `immer` here. This is mainly cause, in some cases, we might make a shallow copy here (e.g. when the error has type `Iterable<ReactNode>`), which might prove to have the same issue we are trying to fix right now.

File: src/components/WebCLI/WebCLI.tsx (hunk 4 above)
Comment (vladimir-cucu) [reply src/components/WebCLI/WebCLI.tsx]: Done.

File: src/components/WebCLI/WebCLI.tsx (hunk 5 above)
Comment (vladimir-cucu) [reply src/components/WebCLI/WebCLI.tsx]: Nice catch! Changed it.

File: src/hooks/useInlineErrors.test.ts (hunk 6 above)
Comment (vladimir-cucu) [reply src/hooks/useInlineErrors.test.ts]: Removed this part.

File: src/hooks/useInlineErrors.test.ts (hunk 7 above)
Comment (vladimir-cucu) [reply src/hooks/useInlineErrors.test.ts]: Removed all the `rerender()` bits.

File: src/components/WebCLI/WebCLI.tsx (hunk 8 above)
Comment (vladimir-cucu) [reply src/components/WebCLI/WebCLI.tsx:178]: Makes sense, added this in my latest commit.

**Review by huwshimi — APPROVED** [review]

**Pull request merged**